src/main/ruby/core/process_mirror.rb

```diff
@@ -173,18 +173,29 @@ def initialize(env_or_cmd, *args)
           end
 
           @options = {}
+          @files_for_child = []
 
           if options
             options.each do |key, value|
               case key
               when ::IO, ::Fixnum, :in, :out, :err
                 from = convert_io_fd key
-                to = convert_to_fd value, from
+                to = convert_to_fd value, from, @files_for_child
                 redirect @options, from, to
               when ::Array
+
+                # We must take care here that we do the right things
+                # when redirecting multiple fds to the same target. We
+                # will do this by handling the first fd normally, and
+                # redirecting the other fds to it. We also require
+                # that the underlying spawn implementation will handle
+                # the initial fd assignment before doing the other fd
+                # redirection.
+
                 from = convert_io_fd key.first
-                to = convert_to_fd value, from
-                key.each { |k| redirect @options, convert_io_fd(k), to }
+                to = convert_to_fd value, from, @files_for_child
+                redirect @options, from, to
+                key[1..-1].each { |k| redirect @options, convert_io_fd(k), from }
```

Comment:
So the other fds are redirected to key.first (from) and not `to`? Why?

Reply:
It was a hangover from an earlier attempt where I let the child open the first fd and just needed to duplicate that for the other fds. Now the child does it all it can be from now.